src/Tools/dotnet-monitor/CollectionRules/CollectionRuleService.cs

```diff
@@ -3,50 +3,38 @@
 // See the LICENSE file in the project root for more information.
 
 using Microsoft.Diagnostics.Monitoring.WebApi;
-using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Actions;
 using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Configuration;
-using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Options;
-using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Triggers;
 using Microsoft.Extensions.Logging;
-using Microsoft.Extensions.Options;
 using System;
 using System.Collections.Generic;
-using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 
 namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules
 {
     internal class CollectionRuleService : IAsyncDisposable
     {
-        private readonly ActionListExecutor _actionListExecutor;
-        private readonly CancellationTokenSource _disposalTokenSource = new();
+        private readonly List<CollectionRuleContainer> _containers = new();
         private readonly ILogger<CollectionRuleService> _logger;
-        private readonly IOptionsMonitor<CollectionRuleOptions> _optionsMonitor;
         private readonly CollectionRulesConfigurationProvider _provider;
-        private readonly List<Task> _runTasks = new();
-        private readonly ICollectionRuleTriggerOperations _triggerOperations;
+        private readonly IServiceProvider _serviceProvider;
 
         private bool _disposed;
 
         public CollectionRuleService(
+            IServiceProvider serviceProvider,
             ILogger<CollectionRuleService> logger,
-            CollectionRulesConfigurationProvider provider,
-            ICollectionRuleTriggerOperations triggerOperations,
-            ActionListExecutor actionListExecutor,
-            IOptionsMonitor<CollectionRuleOptions> optionsMonitor
+            CollectionRulesConfigurationProvider provider
             )
         {
-            _actionListExecutor = actionListExecutor ?? throw new ArgumentNullException(nameof(actionListExecutor));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
             _provider = provider ?? throw new ArgumentNullException(nameof(provider));
-            _optionsMonitor = optionsMonitor ?? throw new ArgumentNullException(nameof(optionsMonitor));
-            _triggerOperations = triggerOperations ?? throw new ArgumentNullException(nameof(triggerOperations));
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
         }
 
         public async ValueTask DisposeAsync()
         {
-            lock (_disposalTokenSource)
+            lock (_containers)
```

Comment:
nit: locking on containers here is a bit misleading since access to containers is otherwise unguarded.

Reply:
Yeah, it's not so much as to protect access to the _containers object, but to prevent simultaneous disposals from attempting to dispose everything.

Reply:
I'm going to remove the lock and flag-check from here and just allow it to dispose each container. The application host is only going to call DisposeAsync once.